### PR TITLE
sync_manifest script

### PR DESCRIPTION
**Pull request by xindiguo**

### CODE REVIEW

**Review by sgosline — APPROVED** [review]
This is a great tool and I wholeheartedly approve it.

**Review by kdaily — COMMENTED** [review]
would suggest a 'local only' that doesn't do anything on Synapse (e.g. shows the manifest that would be created)

File: bin/sync_manifest.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    
+    for index, row in result.iterrows():
+        row[['name','parent']] = _getName(row['path'], synapse_dir, local_root, depth)
+        
+    cols = list(result.columns)
+
+    result = pd.concat([result,pd.DataFrame(columns=key_list)])
+    result = result[cols + key_list] # reorder the columns
+
+    result.to_csv('output.tsv',sep = '\t', index=False)
```
Comment (kdaily) [inline bin/sync_manifest.py]: ~make file name a param or~ write to stdout instead.

File: bin/sync_manifest.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    parser.add_argument('-d','--dir', help='local directory')
+    parser.add_argument('-id','--id',help='Synapse ID of the project/file')
+    parser.add_argument('-f','--files',
+                        help='Path(s) to JSON file(s) of annotations. optional', nargs='+')
+    parser.add_argument('-n','--n', help='depth of hierarchy, DEFAULT is None', default=None)
+    parser.add_argument('-m','--manifest',help='manifest file')
+    args=parser.parse_args()
+    
+    if args.create:
+        sys.stdout.write('Preparing to create manifest\n')
```
Comment (kdaily) [inline bin/sync_manifest.py]: stderr or `logging` package

File: bin/sync_manifest.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            depth = int(depth) 
+
+        dir_list, file_list = _getLists(local_root,depth)
+        print(dir_list)
+        synapse_dir = _getSynapseDir(syn, synapse_id,local_root,dir_list)
+        key_list = _getAnnotationKey(annotations)
+
+        create(file_list,key_list,synapse_dir,local_root,depth)
+    elif args.upload:
+        sys.stdout.write('Preparing to upload files\n')
```
Comment (kdaily) [inline bin/sync_manifest.py]: stderr

File: bin/sync_manifest.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        synu.syncToSynapse(syn, args.manifest)
+    else:
+        sys.stdout.write('Please enter python sync_manifest.py -h for more information.\n')
```
Comment (kdaily) [inline bin/sync_manifest.py]: stderr

**Review by kdaily — COMMENTED** [review]
turn hashtag comments into docstrings

File: bin/sync_manifest.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+def main():
+    import argparse
+    syn = synapseclient.login()
+    
+    parser = argparse.ArgumentParser(description="Create or upload sync manifest")
+    parser.add_argument('-c','--create', help='create manifest', action='store_true')
+    parser.add_argument('-u','--upload', help='sync to synapse using manifest', action='store_true')
+    parser.add_argument('-d','--dir', help='local directory')
+    parser.add_argument('-id','--id',help='Synapse ID of the project/file')
```
Comment (kdaily) [inline bin/sync_manifest.py]: single hyphen options can only be one letter (no `-id`, only `--id`)

**Comment by kdaily** [comment]
This should also be merged to `develop` first, not `master`.

File: bin/sync_manifest.py (hunk 1 above)
Comment (sgosline) [inline bin/sync_manifest.py]: How would that be helpful? Can we make it a default parameter at least?

Comment (kdaily) [inline bin/sync_manifest.py]: It's command line standard operating procedure. I would rather see that you have a script that creates necessary Synapse structure and manifest file, and then let the (soon to be implemented) Synapse sync command line take over (`sync_manifest -c -d mydir/ -i syn123 | synapse sync`). Then it's modular, doesn't rely on changes to Synapse sync command, etc.

File: bin/sync_manifest.py (hunk 2 above)
Comment (xindiguo) [inline bin/sync_manifest.py]: This is only outputting the process. Isn't stderr is for error?

**Pull request merged**